### PR TITLE
Performance Increases

### PR DESCRIPTION
**Pull request by metalshark**
Some minor and some major performance increases. Well worth running a pipe stress test to see the difference.
Changes:
- cache energy/fluid/item handlers and clear caches on invalidation (this required bug fixes to invalidateCaps both here and in other mods) this is by far the biggest performance increase, the PR moved it from ~18% to ~9% CPU usage, this drops it to ~1.5% CPU and decreases disk I/O by requiring 1/8th-1/16th of the getChunk calls on my stress test world
- reduce unnecessary network traffic and actions on the client-side
- prevents the creation of unnecessary objects for comparison and scope objects in lambdas (unless where it aided the JVM, added few places where objects are created where it assisted speed, mainly non-instantly-returned bounding boxes)
- distribute single items in solidifier inventory for use with 3rd party mods interacting (such as RefinedStorage Crafters)
- allow automatically filling tanks above the main tank
- utilise native Forge functionality for extraction filters (isValid), BucketPickupHandlerWrapper, FluidBlockWrapper, FluidUtil.tryFluidTransfer, etc (makes it more likely for re-used code to be "hot" and therefore faster in the JVM)
- allow extraction of fluid from sponges
- fixed infinite fluid sources which regenerated not reappearing after extraction (that built-in BucketPickupHandlerWrapper is magic)
- send repaired input to auto anvils straight to the output slot (if empty) for chaining
- speed and reliability fixes for the Uncrafter (stops getting stuck in edge cases, responds better without having to be emptied, has an output buffer to solve the issue of NOROOM)

### CODE REVIEW

File: src/main/java/com/lothrazar/cyclic/registry/ContainerScreenRegistry.java [hunk 1]
```diff
@@ -62,167 +62,10 @@
 import net.minecraftforge.registries.IForgeRegistry;
 import net.minecraftforge.registries.ObjectHolder;
 
+@SuppressWarnings("checkstyle:StaticVariableName")
```
Comment (metalshark) [inline src/main/java/com/lothrazar/cyclic/registry/ContainerScreenRegistry.java:65]: Added the SuppressWarningsFilter to checkstyle.xml explicitly to use this here.

File: src/main/java/com/lothrazar/cyclic/util/UtilEnergyStorage.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    return LazyOptional.empty();
+  }
+
+  public static int moveEnergy(final IEnergyStorage input, final IEnergyStorage output, final int amount) {
+    int outputCapacity = output.getMaxEnergyStored() - output.getEnergyStored();
+    if (outputCapacity == 0) {
+      return 0;
+    }
+    else if (outputCapacity < 0) {
+      //buggy mods, most likely using a long instead of an int for capacity and it has overflowed
```
Comment (metalshark) [inline src/main/java/com/lothrazar/cyclic/util/UtilEnergyStorage.java:33]: [Industrial Foregoing's Infinity Charger over](https://github.com/InnovativeOnlineIndustries/Industrial-Foregoing/issues/1213)

File: src/main/java/com/lothrazar/cyclic/block/anvil/TileAnvilAuto.java [hunk 3]
```diff
@@ -35,17 +35,26 @@
   static final int MAX = 64000;
   public static IntValue POWERCONF;
   CustomEnergyStorage energy = new CustomEnergyStorage(MAX, MAX);
-  private LazyOptional<IEnergyStorage> energyCap = LazyOptional.of(() -> energy);
+  private final LazyOptional<IEnergyStorage> energyCap = LazyOptional.of(() -> energy);
   ItemStackHandler inputSlots = new ItemStackHandler(1) {
+    @Override
+    public ItemStack insertItem(int slot, ItemStack stack, boolean simulate) {
+      if (!stack.isEmpty() && stack.isRepairable() && stack.getDamage() == 0) {
```
Comment (metalshark) [inline src/main/java/com/lothrazar/cyclic/block/anvil/TileAnvilAuto.java:42]: Shifts repaired items arriving into input straight into output for use in automated lines.

File: src/main/java/com/lothrazar/cyclic/block/solidifier/TileSolidifier.java [hunk 4]
```diff
@@ -173,13 +185,48 @@ private void findMatchingRecipe() {
       return;
     }
     currentRecipe = null;
+    if (!inputSlots.getStackInSlot(0).isEmpty()) {
```
Comment (metalshark) [inline src/main/java/com/lothrazar/cyclic/block/solidifier/TileSolidifier.java:187]: This distributes the items between the slots where possible for helping automation.

File: src/main/java/com/lothrazar/cyclic/block/uncrafter/ContainerUncraft.java [hunk 5]
```diff
@@ -25,7 +25,15 @@ public ContainerUncraft(int windowId, World world, BlockPos pos, PlayerInventory
     ItemStackHandler outputSlots = tile.outputSlots;
     this.endInv = inputSlots.getSlots() + outputSlots.getSlots();
     //the main slot
-    addSlot(new SlotItemHandler(inputSlots, 0, 39, 19));
+    addSlot(new SlotItemHandler(inputSlots, 0, 39, 19) {
+
+      @Override
+      public void onSlotChanged() {
```
Comment (metalshark) [inline src/main/java/com/lothrazar/cyclic/block/uncrafter/ContainerUncraft.java:31]: This took ages to figure out. When a recipe needs multiple of an item, and you right-click to add more, the inventory change methods aren't always called. This forces them to be.

**Comment by Lothrazar** [comment]
there are lot of many really amazing fixes here, but PR is way too big and does to many things all at once. and some style changes are seeming to cause conflicts.  but im making time now to look at everything, as 1.16 i still want to upgrade and move forward as well as porting changes to mc 1.18 after. also saw the kofi <3 so much. ie previous PR on left hand user was perfect size, released that in 1.16 . ill be around every day

**Comment by metalshark** [comment]
Ok understood. Right time for smaller PRs. Didn't realise the checkstyle was causing issues for you? Hopefully would mean it's keeping the existing styling?

**Comment by metalshark** [comment]
First smaller pull request created #2011

**Pull request closed without merging**